### PR TITLE
Repurpose client within `tools/copyblocks` to add `undelete-blocks` and `copyprefix`

### PR DESCRIPTION
**Pull request by andyasp**
#### What this PR does

This PR lifts the object storage client implementation located within `tools/copyblocks` to `pkg/util/objtools` and expands it to include versioning support. The intent is to provide an accessible framework for tools that require more functionality than `objstore` currently supports. Tools are the only target and a lint rule was created to ensure this.

The motivation for this work began with wanting to create an object store generic tool for undelete-blocks-gcs. Restoring an object version requires performing a server-side copy while using an object version as the copy source. `copyblocks` had already implemented server-side copying so it was a nice starting point. Unlike `copyblocks`, the now generic `undelete-blocks` only uses one bucket instead of two. To support this difference `objtools` has both a `BucketConfig` and a `CopyBucketConfig`, where the latter is a composition of the former. I wrote `copyprefix` to experience writing a tool from scratch with `CopyBucketConfig`. It helped with the design and made separating code from `copyblocks` easier. An effort has also been made to align with the flags used in `objstore` to reduce the cognitive burden incurred when switching between tools.

The code is rebased into 3 commits:
- Migrating the client from `copyblocks` to `pkg/util/objtools` and expanding its functionality
- Adding the generic `undelete-blocks` tool
- Adding the generic `copyprefix` tool

Needless to say, this is a large PR and I'm happy with separating this if needed. I opened this as one to make it easier to see the reasoning behind some of the choices made.

As far as testing goes, it's still manual (aided by scripts). It'd be nice for that to change, but this was already quite a bit of work.

## copyblocks

Note: `--dry-run` only since `copyprefix` exercises the same client code paths with easier setup.

Tests Performed:
- [x] ABS -> GCS
- [x] ABS -> S3
- [x] GCS -> ABS 
- [x] GCS -> S3 
- [x] S3  -> ABS
- [x] S3  -> GCS 

## undelete-blocks

Note: Partial block deletions, recovery, followed by whole block deletions. Ensured same block can be recovered more than once.

Tests Performed:
- [x] S3
- [x] ABS
- [x] GCS

## copyprefix

Note: all copies were performed across buckets rather than only on the same one since `undelete-blocks` exercises same bucket copying.

Tests Performed:
- [x] ABS -> ABS
- [x] ABS -> ABS (client-side)
- [x] ABS -> GCS
- [x] ABS -> S3
- [x] GCS -> ABS 
- [x] GCS -> GCS 
- [x] GCS -> GCS (client-side)
- [x] GCS -> S3 
- [x] S3  -> ABS
- [x] S3  -> GCS 
- [x] S3  -> S3
- [x] S3  -> S3 (client-side)

#### Which issue(s) this PR fixes or relates to

Fixes N/A

#### Checklist

- [ ] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: tools/copyblocks/main.go
Comment (andyasp) [inline tools/copyblocks/main.go:336]: This seems like the added `meta.json` could have a bug in regards to what has a prefix and what doesn't. I'll double check this and do a non dry-run after to confirm.

Comment (andyasp) [reply tools/copyblocks/main.go]: Looked like there were two bugs here actually, the prefix bug and a double counting bug. Rewrote it to simplify it and did a test run that went well.

**Review by pstibrany — COMMENTED** [review]
"undelete-blocks" tool is doing too much. I would strongly prefer to split its functionality between 1) listing deleted blocks for a single tenant, 2) undeleting list of blocks for single tenant. I would completely remove option to work with JSON file input.

Furthermore I would suggest to split this PR into
1) tool to refactor bucket code out of copyblocks 
2) PR for copyprefix
3) PR for listing deleted blocks, and undeleting individual blocks

**Comment by andyasp** [comment]
Thanks for taking a look at this @pstibrany

> "undelete-blocks" tool is doing too much. I would strongly prefer to split its functionality between 1) listing deleted blocks for a single tenant, 2) undeleting list of blocks for single tenant. I would completely remove option to work with JSON file input.

Respectfully disagree, let's talk through each point.

> "undelete-blocks" tool is doing too much.

It undeletes blocks. It needs to know what to undelete in order to do this and two options are provided: either every block for tenants or through an exact list of blocks. The former is a very useful default (especially in stressful situations) and the latter is an escape hatch for avoiding complexity within the tool itself.

> 1) listing deleted blocks for a single tenant

`listblocks` already exists, and even contains a `showDeleted` option among others that pair well with this. It does not currently have the ability to list fully deleted blocks (from only noncurrent versions) because it didn't have the ability to list versions. That could potentially be an extension to it (currently that would be at the cost of auth options `objstore` supports and `objtools` does not), but I don't think a new tool should be added for that functionality.

> 2) undeleting list of blocks for single tenant

I'm doing basically what `copyblocks` already does. I'd rather a user can say `--include-tenants tenant1,tenant2` than have to write a for loop in bash to do the same. With excluded tenants it's even more bash work. With the utility you pointed to (thank you, I didn't even know that existed) it's basically only a for loop over tenants internally.

> I would completely remove option to work with JSON file input.

The point of the JSON input is to avoid complexity. Let's say someone wiped their blocks bucket but actually wanted some blocks back that met some arbitrary criteria. I don't want to support selection criteria at all in this tool and don't have to. They can use whatever else to get the required information, provide it in a widely used format and then be done. I made it JSON to provide something more standard than the line-based formatting accepted in `undelete-blocks-gcs`.

> Furthermore I would suggest to split this PR

The rebased commits are for that if needed. My one worry is if unused code from the first refactor commit gets complex, but that can probably be worked around.

**Comment by pstibrany** [comment]
> > "undelete-blocks" tool is doing too much.
> 
> It undeletes blocks. It needs to know what to undelete in order to do this and two options are provided: either every block for tenants or through an exact list of blocks. The former is a very useful default (especially in stressful situations) and the latter is an escape hatch for avoiding complexity within the tool itself.

It not only undeletes blocks, but also lists "fully deleted" blocks. This is useful functionality on its own, and it would be good to have ability to run command to only list such blocks, for further inspection and modification before passing it to undelete. I'm basically proposing to split "undelete-blocks" into two useful tools, that can be combined together.

> > 1. listing deleted blocks for a single tenant
> 
> `listblocks` already exists, and even contains a `showDeleted` option among others that pair well with this. It does not currently have the ability to list fully deleted blocks

Exactly. "listblocks" is different in that it loads meta.json files. I wouldn't mix "listing fully deleted blocks" with that functionality. Furthermore listblocks is also part of store-gateways' HTTP API, I don't think we want to modify store-gateways to work with new bucket code.

> > 2. undeleting list of blocks for single tenant
> 
> I'm doing basically what `copyblocks` already does. I'd rather a user can say `--include-tenants tenant1,tenant2` than have to write a for loop in bash to do the same. With excluded tenants it's even more bash work. With the utility you pointed to (thank you, I didn't even know that existed) it's basically only a for loop over tenants internally.

`copyblocks` is designed to run as a long-running service, that periodically copies all blocks between buckets.

"undelete-blocks" is command line utility. If we make it simpler to work with, it will be easier to script.

> > I would completely remove option to work with JSON file input.
> 
> The point of the JSON input is to avoid complexity. Let's say someone wiped their blocks bucket but actually wanted some blocks back that met some arbitrary criteria. I don't want to support selection criteria at all in this tool and don't have to. They can use whatever else to get the required information, provide it in a widely used format and then be done. I made it JSON to provide something more standard than the line-based formatting accepted in `undelete-blocks-gcs`.

I agree we don't want any complexity in "undelete blocks"... it should basically work as "undelete-blocks-gcs" does today. (I misremembered undelete-blocks-gcs using command line arguments and works for single tenant, but that's not what it does. However I would still argue that one-block per line format is easier to generate in scripts than JSON).

"list deleted blocks" tool should generate input for "undelete blocks".

> > Furthermore I would suggest to split this PR
> 
> The rebased commits are for that if needed. My one worry is if unused code from the first refactor commit gets complex, but that can probably be worked around.

Separate PRs would be faster to review. Right now this PR does refactoring, it deletes existing tool, it adds two new tools (possibly three if we split undelete blocks).

**Comment by andyasp** [comment]
> It not only undeletes blocks, but also lists "fully deleted" blocks. This is useful functionality on its own, and it would be good to have ability to run command to only list such blocks, for further inspection and modification before passing it to undelete. I'm basically proposing to split "undelete-blocks" into two useful tools, that can be combined together.


If you want an answer to "which of my blocks are deleted/recoverable, and what actions would be performed to do it?" that is already satisfied here with `--dry-run`. It's using structured logging very intentionally. I don't understand how that's a downside. If you feed it blocks it tells you the results. If you don't then it looks at all the blocks. It helped immensely when I was testing this. The last thing I want to do in an incident is write scripts on the fly when there are predictable targets.
> copyblocks is designed to run as a long-running service, that periodically copies all blocks between buckets.
"undelete-blocks" is command line utility. If we make it simpler to work with, it will be easier to script.

It's easier to not have to write a script at all and get what you need. That's not to say that you couldn't use another tool and pair it with this if you needed to, that's exactly what the JSON input is for.

I think you're seeing value in the block listing with versions here and I agree with you, but I intentionally made this pass at it very straightforward. with I agree with, but we're thinking about it end to end in separate ways. To me `listblocks` is about listing blocks. The fact that it shares anything with the store gateway is incidental. It supports `--min-time` and `--max-time` already. I don't see value in separating listing partially or fully deleted blocks into its own tool in the way that you're proposing when that could be a field in the output format. When writing this I had even considered trying to add a `--output-format=json` to `listblocks` as a first step (I didn't, this was already a lot).


> Exactly. "listblocks" is different in that it loads meta.json files. I wouldn't mix "listing fully deleted blocks" with that functionality. Furthermore listblocks is also part of store-gateways' HTTP API, I don't think we want to modify store-gateways to work with new bucket code.

This loads `meta.json` files. It uses it to identify block objects then makes sure each has a recoverable version before restoring anything.

**Comment by andyasp** [comment]
Sorry if you got two pings for this, was drafting a response and accidentally hit Shift+Enter then had to start over from scratch.
 
> It not only undeletes blocks, but also lists "fully deleted" blocks. This is useful functionality on its own, and it would be good to have ability to run command to only list such blocks, for further inspection and modification before passing it to undelete. I'm basically proposing to split "undelete-blocks" into two useful tools, that can be combined together.

I agree that listing blocks with versioning on its own is valuable, but my viewpoint is that the listing tool should be `listblocks`. Whether `listblocks` shares code with the store gateway currently is incidental and a matter of code organization. I almost even wrote an `--output-format=json` for `listblocks`, since eventually if it supported versioned listing it could have a field like `IsDeleted`. It could be super nice to pair that with `--min-time` or `--max-time`. I'm biased towards liking `jq` though.

I disagree that the above means that this tool should not have simplified listing functionality. The users of this tool likely have predictable targets like an entire bucket or a tenant. There's immense value in having common tasks be supported without additional scripting, especially in an incident scenario with time pressure. The `--dry-run` option paired with structured logging here is very powerful as is and might even satisfy being the listing tool you're asking for (try it!). If the preferred blocks aren't predictable then another tool can be used with the JSON input, that's why it was written.

> copyblocks is designed to run as a long-running service, that periodically copies all blocks between buckets.
"undelete-blocks" is command line utility. If we make it simpler to work with, it will be easier to script.

It's easier to not have to script at all if you don't have to. `copyblocks` is similar because it's working on the same block level as `undelete-blocks`. The code practically gives away that I wrote `undelete-blocks` after staring a lot at `copyblocks`. It has tenant options just like this does. This tool might also take a while if it has to recover an entire bucket. I predicted criticism for not including concurrency options for tenant/blocks (which I left out for simplicity on the first pass).

> I agree we don't want any complexity in "undelete blocks"... it should basically work as "undelete-blocks-gcs" does today. I misremembered undelete-blocks-gcs using command line arguments and works for single tenant, but that's not what it does. 

I think where we're diverging is you are thinking of the complexity from the code perspective and I'm thinking of it from a user experience perspective. I can envision a future where versioned tenant/block listing is somewhere in `util` where both `listblocks` and `undelete-blocks` could use it and then we could both be happy. Not proposing doing that right now though since this work is focused on getting code out of `copyblocks` and into `util`.

> However I would still argue that one-block per line format is easier to generate in scripts than JSON). "list deleted blocks" tool should generate input for "undelete blocks".

JSON is far easier to grok and generate to me because it's relatively standard.

> Separate PRs would be faster to review. Right now this PR does refactoring, it deletes existing tool, it adds two new tools (possibly three if we split undelete blocks).

Oh yeah, I know this is a lot. I expected a discussion and wanted to be clear about my intentions before introducing the scaffolding.

File: tools/undelete-blocks/main.go
Comment (pstibrany) [inline tools/undelete-blocks/main.go:121]: Let's get blocks from stdin, by reading paths to the blocks. I don't think we need to support JSON vs lines, or input file. (stdin works fine).

File: tools/undelete-blocks/main.go
Comment (pstibrany) [inline tools/undelete-blocks/main.go:39]: I would remove tenant filtering functionality. This tool should restore all supplied blocks.

File: tools/undelete-blocks/main.go
Comment (pstibrany) [inline tools/undelete-blocks/main.go:191]: I am really uncomfortable by letting the tool to list tenants and doing the undelete based on filters. Let's move listing tenants and blocks to separate tool. Then we can inspect the output of listing before deciding how to proceed.

File: tools/undelete-block-gcs/main.go
Comment (pstibrany) [inline tools/undelete-block-gcs/main.go:1]: I don't think we should delete this tool just yet. I think it's still useful, and much simpler to work with.

File: tools/undelete-blocks/main.go
Comment (pstibrany) [reply tools/undelete-blocks/main.go:381]: I was thinking about "Not found" error. Other errors should be reported.

File: tools/undelete-blocks/main.go
Comment (andyasp) [reply tools/undelete-blocks/main.go]: Aligned to this except when a version restore or a delete marker delete fails since the structured logging is useful in those cases.

File: tools/undelete-blocks/main.go
Comment (andyasp) [reply tools/undelete-blocks/main.go]: I have no preference between space or `/`. I swapped it to `/`. Are you saying you want support for a trailing delimiter?

File: tools/undelete-blocks/main.go
Comment (andyasp) [reply tools/undelete-blocks/main.go]: > Let's get blocks from stdin, by reading paths to the blocks.

Sure, to do this I've added `--blocks-from`, which gives the choices `json`, `lines`, and `listing`. If `--input-file` is not supplied with `json` or `lines` it now defaults to reading from stdin. The net effect is you can say 
```bash
echo "tenant/01GVC9CD6KGMNPVB47X9S208NF" | ./undelete-blocks --blocks-from lines
```

> I don't think we need to support JSON vs lines

We've talked about this externally, but for posterity the reason I want JSON support was to provide a pathway for selecting on multiple block attributes with something like `jq`.

File: tools/undelete-blocks/main.go
Comment (andyasp) [reply tools/undelete-blocks/main.go:132]: Answered above.

File: tools/undelete-blocks/main.go
Comment (andyasp) [reply tools/undelete-blocks/main.go]: The tenant filtering was very intentionally added and has a purpose. For the listing option it gives targeting. For the `lines` and `json` options it allows a large input to be reused without externally modifying it.

File: tools/undelete-block-gcs/main.go
Comment (andyasp) [reply tools/undelete-block-gcs/main.go:1]: > I don't think we should delete this tool just yet. I think it's still useful, and much simpler to work with.

If you apply equal standards to each tool then this doesn't make sense. Not dealing with complexity doesn't make it go away. The GCS tool has a number of drawbacks that this tool does not:

-  Despite the name being `undelete-block-gcs` there is nothing tying the input to be on block prefixes. This is not a virtue since its entirely on the user to prepare the undocumented input format.
-  It does not handle multiple versions of an object gracefully. If there is a noncurrent version that exists, it restores it. Even if a current version exists for the same object it does this. When there are multiple noncurrent versions for the same object it does this repeatedly, overwriting itself and generating even more noncurrent versions. In particular this is a barrier to reruns with the same input if an error is encountered.
- It will restore any object (besides a delete marker) with a noncurrent version within the block prefix, even when restoring the block fully is impossible. If a chunk object had no noncurrent versions that restored block is now corrupt. If a block had been backfilled it would restore the temporary objects backfill adds.
- Deleting delete markers (both local and global) is unhandled. It assumes the block is fully deleted.

Despite the above, the GCS tool did its job when it needed to very successfully. What I'm trying to do here is iterate and solve these problems to make it more accessible. It's more lines of code, sure, but there's a reason for it and I want us to align on that.

File: tools/undelete-blocks/main.go
Comment (andyasp) [reply tools/undelete-blocks/main.go:199]: I removed the listing behavior as the default and forced the user to provide `--blocks-from listing` as an attempt to address your safety concerns. We've talked about adding support for `listblocks` for more precise filtering.

File: tools/undelete-blocks/main.go
Comment (andyasp) [inline tools/undelete-blocks/main.go:328]: This `version` struct and the `restorableVersion` struct don't add much beyond not leaking the `objtools.ObjectAttributes` type too far into the restore logic.  A single struct could probably achieve that same goal more cleanly.

File: tools/undelete-blocks/main.go
Comment (andyasp) [reply tools/undelete-blocks/main.go:87]: considering this. If we wanted to bypass the tenant listing on `--blocks-from listing` with a `--include-tenants` it would have to be extended. might be better to try that out locally and then see if we wanted to add that to the `util` version.

**Comment by leizor** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

File: tools/undelete-blocks/README.md
Comment (pstibrany) [inline tools/undelete-blocks/README.md:46]: lol, I don't think people will need to translate one format to the other :) We can drop this.

File: tools/undelete-blocks/README.md
Comment (pstibrany) [inline tools/undelete-blocks/README.md:57]: "Nothing" is not a write operation. We can simply say that when block is uploaded to the storage, `meta.json` file is uploaded last.

File: tools/undelete-blocks/README.md
Comment (pstibrany) [inline tools/undelete-blocks/README.md:64]: note: When compactor decides which blocks to compact, it skips blocks based on global markers files. It ignores "local" marker file.

(it's just clarification, no need to to update text)

File: tools/undelete-blocks/README.md
Comment (pstibrany) [inline tools/undelete-blocks/README.md:65]: what is meant by "noticed" is to include deletion markers in tenant's bucket-index, and let store-gateways to unload the blocks.

File: tools/undelete-blocks/README.md
Comment (pstibrany) [inline tools/undelete-blocks/README.md:73]: > For instance, if a `no-compact-mark.json` was present it will not be restored by this tool.

Hmm, this is interesting choice. If there was `no-compact-mark.json` file, it was likely that block was causing troubles during compaction. By not restoring that file, we can cause troubles again.

(Compactor actually ignores "local" no-compact mark file, and only checks for global one)

I'd suggest to improve the tool to restore this file too. (Not necessarily in this PR)

File: tools/undelete-blocks/README.md
Comment (pstibrany) [inline tools/undelete-blocks/README.md:77]: I think this tool is useful enough to include it in released binaries. (Can be done in separate PR).

File: tools/undelete-blocks/README.md
Comment (pstibrany) [inline tools/undelete-blocks/README.md:94]: Don't we call this "azure" everywhere else?

File: CHANGELOG.md
Comment (pstibrany) [inline CHANGELOG.md:126]: Please don't forget to rebase and move these entries to `main` again.

File: tools/undelete-blocks/main.go
Comment (andyasp) [reply tools/undelete-blocks/main.go]: Fully aligned to this now after a bit of a refactor.

File: tools/undelete-blocks/README.md
Comment (andyasp) [reply tools/undelete-blocks/README.md]: haha but it was so fun to write! Fair point, I removed it. This was an attempt to bridge the input preferences if there was ever external tooling written around it, but that is a little much.

File: tools/undelete-blocks/README.md
Comment (andyasp) [reply tools/undelete-blocks/README.md]: Removed the nothing.

File: tools/undelete-blocks/README.md
Comment (andyasp) [reply tools/undelete-blocks/README.md:71]: That'd be nice! I remember a previous discussion about trying to fit some tools into `mimirtool`.

File: tools/undelete-blocks/README.md
Comment (andyasp) [reply tools/undelete-blocks/README.md]: Nice catch! This was leftover from before the flags were aligned. I corrected it here and in the `copyblocks` README as well.

File: CHANGELOG.md
Comment (andyasp) [reply CHANGELOG.md:126]: :+1: I'll be sure to do this before I do the last round of manual testing for the new changes. I marked this as draft until that's completed.

File: tools/undelete-blocks/main.go
Comment (andyasp) [reply tools/undelete-blocks/main.go]: Nice, changed it to `-`.

**Comment by andyasp** [comment]
As a note I added size enforcement according to `meta.json` for picking versions in the last round of changes. The particular case I'm adding this for is an Azure case where if the server-side copy is [pending or fails it can leave around a 0-length object](https://learn.microsoft.com/en-us/rest/api/storageservices/copy-blob?tabs=microsoft-entra-id#working-with-a-pending-copy-operation-version-2012-02-12-and-later). With no size enforcement this object would inhibit the tool from picking a past version and completing the restore. This seemed cleaner than special casing `0` somehow or trying to detect it in the listing, but does add some more reliance on the `meta.json` being filled out. I'll see from internal testing if it behaves okay.

File: tools/undelete-blocks/README.md
Comment (andyasp) [reply tools/undelete-blocks/README.md]: I tried to make it clearer by changing it to:
`
3. Some time passes to provide time for the delete markers to be picked up and added to the bucket index.
`

**Comment by andyasp** [comment]
Retested `undelete-blocks` on azure, s3, and gcs and didn't run into any surprises beyond a silly refactoring mistake. Off the top of my head the work left to do for `undelete-blocks` that was discussed:
- Stop listing tenants when `--include-tenants` is passed with `--blocks-from=list`
- Add concurrency options
- Enhance `listblocks` to do versioned listing with the goal of generating useful input for `undelete-blocks` in some way
- Add it to the released binaries or to mimirtool to make it more accessible

Those all seem better done in their own PRs. @pstibrany I'll wait for your final go ahead to merge once you're back since the checked size and no-compact marker changes went in after your approval.

File: tools/undelete-blocks/README.md
Comment (pstibrany) [reply tools/undelete-blocks/README.md:71]: > That'd be nice! I remember a previous discussion about trying to fit some tools into `mimirtool`.

Eventually, we may want to do that. For now, standalone tool is fine too.

**Review by pstibrany — APPROVED** [review]
Thank you, great work!

**Pull request merged**
